server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/cs/deployment/AbstractCacheStoreExtensionProcessor.java

```diff
@@ -47,55 +47,60 @@ private void addServices(DeploymentPhaseContext ctx, ServicesAttachment services
+      AbstractExtensionManagerService<T> service = createService(implementationClassName, instanceFactory);
       InfinispanLogger.ROOT_LOGGER.installDeployedCacheStore(implementationClassName);
-      ServiceBuilder<T> serviceBuilder = ctx.getServiceTarget().addService(extensionServiceName, service);
-      serviceBuilder.setInitialMode(ServiceController.Mode.ACTIVE);
-      serviceBuilder.addDependency(DeployedCacheStoreFactoryService.SERVICE_NAME, DeployedCacheStoreFactory.class, service.getDeployedCacheStoreFactory());
+
+      ServiceBuilder<?> serviceBuilder = ctx.getServiceTarget().addService(service.getName()).setInitialMode(ServiceController.Mode.ACTIVE);
+      service.deployedCacheStoreFactory = serviceBuilder.requires(DeployedCacheStoreFactoryService.SERVICE_NAME);
+      serviceBuilder.setInstance(service);
       serviceBuilder.install();
    }
 
    public abstract Class<T> getServiceClass();
 
-   public abstract AbstractExtensionManagerService<T> createService(String serviceName, T instance);
+   public abstract AbstractExtensionManagerService<T> createService(String implClassName, Supplier<T> instanceFactory);
 
-   protected static abstract class AbstractExtensionManagerService<T> implements Service<T> {
+   static abstract class AbstractExtensionManagerService<T> implements Service {
 
-      protected final T extension;
-      protected final String className;
-      protected final InjectedValue<DeployedCacheStoreFactory> deployedCacheStoreFactory = new InjectedValue<>();
+      final Supplier<T> instanceFactory;
+      final String className;
+      final String serviceName;
+      Supplier<DeployedCacheStoreFactory> deployedCacheStoreFactory;
 
-      protected AbstractExtensionManagerService(String className, T extension) {
-         this.extension = extension;
+      AbstractExtensionManagerService(String serviceName, String className, Supplier<T> instanceFactory) {
+         this.serviceName = serviceName;
          this.className = className;
+         this.instanceFactory = instanceFactory;
       }
 
       @Override
       public void start(StartContext context) {
          InfinispanLogger.ROOT_LOGGER.deployedStoreStarted(className);
-         deployedCacheStoreFactory.getValue().addInstance(extension);
+         deployedCacheStoreFactory.get().addInstanceFactory(className, instanceFactory);
       }
 
       @Override
       public void stop(StopContext context) {
          InfinispanLogger.ROOT_LOGGER.deployedStoreStopped(className);
-         deployedCacheStoreFactory.getValue().removeInstance(extension);
+         deployedCacheStoreFactory.get().removeInstance(className);
       }
 
-      InjectedValue<DeployedCacheStoreFactory> getDeployedCacheStoreFactory() {
-         return deployedCacheStoreFactory;
+      public ServiceName getName() {
+         return ServiceName.JBOSS.append(serviceName, className.replaceAll("\\.", "_"));
```

Comment:
When replacing a single char it is better to use public String replace(char oldChar, char newChar)